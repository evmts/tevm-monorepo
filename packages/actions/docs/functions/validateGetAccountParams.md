[**@tevm/actions**](../README.md)

***

[@tevm/actions](../globals.md) / validateGetAccountParams

# Function: validateGetAccountParams()

> **validateGetAccountParams**(`action`): `InvalidRequestError`[]

Defined in: [packages/actions/src/GetAccount/validateGetAccountParams.js:12](https://github.com/evmts/tevm-monorepo/blob/main/packages/actions/src/GetAccount/validateGetAccountParams.js#L12)

## Parameters

### action

[`GetAccountParams`](../type-aliases/GetAccountParams.md)\<`boolean`\>

## Returns

`InvalidRequestError`[]
