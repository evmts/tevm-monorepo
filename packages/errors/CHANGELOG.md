# @tevm/errors

## 2.0.0-next.105

### Minor Changes

- [#1370](https://github.com/evmts/tevm-monorepo/pull/1370) [`1dcfd69`](https://github.com/evmts/tevm-monorepo/commit/1dcfd6944f77493a00daa0d64590c2b0c0983a0f) Thanks [@roninjin10](https://github.com/roninjin10)! - Renamed tevm/base-client to tevm/node

## 1.1.0-next.100

### Patch Changes

- [#1322](https://github.com/evmts/tevm-monorepo/pull/1322) [`6407be7`](https://github.com/evmts/tevm-monorepo/commit/6407be7736c996aa8939a0ec5ee13c3d3c34f1e5) Thanks [@roninjin10](https://github.com/roninjin10)! - Migrated to vitest for better coverage reporting

## 1.1.0-next.96

### Patch Changes

- [#1301](https://github.com/evmts/tevm-monorepo/pull/1301) [`59268b2`](https://github.com/evmts/tevm-monorepo/commit/59268b2e00423ba8f9ddf6fa89ea0070ae1023a6) Thanks [@roninjin10](https://github.com/roninjin10)! - Added sideEffect: false to package.json for better tree shaking support

## 2.0.0-next.86

### Patch Changes

- [#1240](https://github.com/evmts/tevm-monorepo/pull/1240) [`b1f0504`](https://github.com/evmts/tevm-monorepo/commit/b1f0504a6be9abf08faa029e7b40fb53666293e8) Thanks [@roninjin10](https://github.com/roninjin10)! - Bumped sub dep up

## 2.0.0-next.84

### Patch Changes

- [#1232](https://github.com/evmts/tevm-monorepo/pull/1232) [`a170f0f`](https://github.com/evmts/tevm-monorepo/commit/a170f0f05a624f70cadea95f4fbaf11c00d5cadd) Thanks [@roninjin10](https://github.com/roninjin10)! - Fix bug with error tags not being set properly

- [#1232](https://github.com/evmts/tevm-monorepo/pull/1232) [`a170f0f`](https://github.com/evmts/tevm-monorepo/commit/a170f0f05a624f70cadea95f4fbaf11c00d5cadd) Thanks [@roninjin10](https://github.com/roninjin10)! - Fixed issues with error handling and added unit testing

## 2.0.0-next.80

### Patch Changes

- [#1221](https://github.com/evmts/tevm-monorepo/pull/1221) [`b0b63d2`](https://github.com/evmts/tevm-monorepo/commit/b0b63d22076f35d76898ab1094ece9668ceef95d) Thanks [@roninjin10](https://github.com/roninjin10)! - Bump bundler

## 2.0.0-next.79

### Minor Changes

- [#1210](https://github.com/evmts/tevm-monorepo/pull/1210) [`f2d4ac4`](https://github.com/evmts/tevm-monorepo/commit/f2d4ac43dab0c5affe994985851639438cb05911) Thanks [@roninjin10](https://github.com/roninjin10)! - Add compatability for viem code property

### Patch Changes

- [#1211](https://github.com/evmts/tevm-monorepo/pull/1211) [`f51ef40`](https://github.com/evmts/tevm-monorepo/commit/f51ef4007f53b2ca6d4ebff770104d9e9f462ea3) Thanks [@roninjin10](https://github.com/roninjin10)! - Update all packages for new tevm contract changes"

- [#1219](https://github.com/evmts/tevm-monorepo/pull/1219) [`a8070b7`](https://github.com/evmts/tevm-monorepo/commit/a8070b769da6695d5e27569809f8ac86866b081d) Thanks [@roninjin10](https://github.com/roninjin10)! - Making sure every package releases

- [#1213](https://github.com/evmts/tevm-monorepo/pull/1213) [`84a6d9c`](https://github.com/evmts/tevm-monorepo/commit/84a6d9caae5e72246933d72e8721d466b238cf81) Thanks [@roninjin10](https://github.com/roninjin10)! - Upgraded all dependencies except effect to latest

## 1.1.0-next.77

### Patch Changes

- [#1200](https://github.com/evmts/tevm-monorepo/pull/1200) [`398daa0`](https://github.com/evmts/tevm-monorepo/commit/398daa059ed1c4373200da1a114ef07d156b207d) Thanks [@roninjin10](https://github.com/roninjin10)! - Rerelease tevm packages

## 1.1.0-next.75

### Patch Changes

- [#1193](https://github.com/evmts/tevm-monorepo/pull/1193) [`db7bfc7`](https://github.com/evmts/tevm-monorepo/commit/db7bfc7bac341e29e2df20569347eb019e2d37a7) Thanks [@roninjin10](https://github.com/roninjin10)! - Updated server, utils and trie packages. Update readmes. Update jsdoc. Update error handling. Fix corner case bugs found. Add test coverage

## 1.1.0-next.74

### Minor Changes

- [#1186](https://github.com/evmts/tevm-monorepo/pull/1186) [`7765446`](https://github.com/evmts/tevm-monorepo/commit/7765446beec1391a00f3d3dd8d015d5205e0371a) Thanks [@roninjin10](https://github.com/roninjin10)! - Implemented new typesafe error system based on viem errors and the JSON-RPC spec for error codes. All errors come with a typesafe `name` property along with link to docs about the error. They also have a `code` property that maps to ethereum JSON-RPC error codes. All concrete errors are implemented in the `@tevm/errors` package. Each function will then export a union error type of all the errors it and it's sub-functions can throw.

- [#1186](https://github.com/evmts/tevm-monorepo/pull/1186) [`7765446`](https://github.com/evmts/tevm-monorepo/commit/7765446beec1391a00f3d3dd8d015d5205e0371a) Thanks [@roninjin10](https://github.com/roninjin10)! - Renamed action errors like CallError to TevmCallError. They are exported from actions package now instead of errors package

### Patch Changes

- [#1186](https://github.com/evmts/tevm-monorepo/pull/1186) [`7765446`](https://github.com/evmts/tevm-monorepo/commit/7765446beec1391a00f3d3dd8d015d5205e0371a) Thanks [@roninjin10](https://github.com/roninjin10)! - Moved files around to colocate code better. Some packages are disappearing

  - Tevm/Zod is now part of Tevm/actions
  - Tevm/actions-types moved to Tevm/actions
  - Tevm/procedures-types moved to Tevm/procedures

## 1.1.0-next.72

### Minor Changes

- [`4094ead`](https://github.com/evmts/tevm-monorepo/commit/4094eadc105790d4e4046187772a8cdbf28c0ef9) - Fix changesets

## 1.1.0-next.69

### Patch Changes

- [#1163](https://github.com/evmts/tevm-monorepo/pull/1163) [`eb80103`](https://github.com/evmts/tevm-monorepo/commit/eb80103442991a0bbe8342f1e237cb52b1a8cc9e) Thanks [@roninjin10](https://github.com/roninjin10)! - Fix bad publish

## 1.1.0-next.60

### Patch Changes

- [#1127](https://github.com/evmts/tevm-monorepo/pull/1127) [`1314a07`](https://github.com/evmts/tevm-monorepo/commit/1314a0770007dd3aa8a4762ddbec62ac60c1dfb6) Thanks [@roninjin10](https://github.com/roninjin10)! - Fixed bad release from lack of changeset

## 1.1.0-next.52

### Patch Changes

- [#1088](https://github.com/evmts/tevm-monorepo/pull/1088) [`65e4089`](https://github.com/evmts/tevm-monorepo/commit/65e40891fca12fc7fde5d1e177527cd70b28cb1f) Thanks [@roninjin10](https://github.com/roninjin10)! - Fixed bug with failing to include changeset for tx package. Bumping every package just to be safe

## 1.1.0-next.47

### Patch Changes

- [#1064](https://github.com/evmts/tevm-monorepo/pull/1064) [`85d9143`](https://github.com/evmts/tevm-monorepo/commit/85d9143262396416a5d94b527f7bd3148ab51069) Thanks [@roninjin10](https://github.com/roninjin10)! - Fixed test-utils package being mistakedly private

## 1.1.0-next.45

### Minor Changes

- [#1042](https://github.com/evmts/tevm-monorepo/pull/1042) [`2a00b2f`](https://github.com/evmts/tevm-monorepo/commit/2a00b2fe10171aaa0607aed66c29d8df8c3437c8) Thanks [@roninjin10](https://github.com/roninjin10)! - Added new tevm_mine and anvil_mine support for mining blocks. This includes the JSON-RPC support as well as a new tevm.mine() action to the main tevm api

## 1.0.0-next.40

### Patch Changes

- [#962](https://github.com/evmts/tevm-monorepo/pull/962) [`64c3c6dd09c236c516b117a506380028c3154c4a`](https://github.com/evmts/tevm-monorepo/commit/64c3c6dd09c236c516b117a506380028c3154c4a) Thanks [@roninjin10](https://github.com/roninjin10)! - Added docs for all packages to https://tevm.sh

## 1.0.0-next.28

### Patch Changes

- [#913](https://github.com/evmts/tevm-monorepo/pull/913) [`06268ffeebaec950d1606732c4eb6fd5f9bb77d5`](https://github.com/evmts/tevm-monorepo/commit/06268ffeebaec950d1606732c4eb6fd5f9bb77d5) Thanks [@roninjin10](https://github.com/roninjin10)! - Fixed bug with packages linking to older versions of tevm

## 1.0.0-next.23

### Patch Changes

- [#862](https://github.com/evmts/tevm-monorepo/pull/862) [`f217fa4dc2f730cc109940ef36872229ae9f26d3`](https://github.com/evmts/tevm-monorepo/commit/f217fa4dc2f730cc109940ef36872229ae9f26d3) Thanks [@roninjin10](https://github.com/roninjin10)! - - Renamed MemoryTevm MemoryClient
  - Renamed TevmClient HttpClient
  - Replaced @tevm/actions-types package with @tevm/actions-types, @tevm/client-types, and @tevm/procedures-types packages
  - Moved errors to @tevm/errors
  - Moved bundler packages out of tevm and to @tevm/bundler package
  - Minimized packages exposed in tevm package
  - Fixed bug with missing types exports
