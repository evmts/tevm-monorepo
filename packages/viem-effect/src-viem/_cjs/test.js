'use strict'
Object.defineProperty(exports, '__esModule', { value: true })
exports.stopImpersonatingAccount =
	exports.setRpcUrl =
	exports.snapshot =
	exports.setStorageAt =
	exports.setNonce =
	exports.setNextBlockTimestamp =
	exports.setNextBlockBaseFeePerGas =
	exports.setMinGasPrice =
	exports.setLoggingEnabled =
	exports.setIntervalMining =
	exports.setCoinbase =
	exports.setCode =
	exports.setBlockTimestampInterval =
	exports.setBlockGasLimit =
	exports.setBalance =
	exports.setAutomine =
	exports.sendUnsignedTransaction =
	exports.revert =
	exports.reset =
	exports.removeBlockTimestampInterval =
	exports.mine =
	exports.inspectTxpool =
	exports.increaseTime =
	exports.impersonateAccount =
	exports.getTxpoolStatus =
	exports.getTxpoolContent =
	exports.getAutomine =
	exports.dropTransaction =
		void 0
const dropTransaction_js_1 = require('./actions/test/dropTransaction.js')
Object.defineProperty(exports, 'dropTransaction', {
	enumerable: true,
	get: function () {
		return dropTransaction_js_1.dropTransaction
	},
})
const getAutomine_js_1 = require('./actions/test/getAutomine.js')
Object.defineProperty(exports, 'getAutomine', {
	enumerable: true,
	get: function () {
		return getAutomine_js_1.getAutomine
	},
})
const getTxpoolContent_js_1 = require('./actions/test/getTxpoolContent.js')
Object.defineProperty(exports, 'getTxpoolContent', {
	enumerable: true,
	get: function () {
		return getTxpoolContent_js_1.getTxpoolContent
	},
})
const getTxpoolStatus_js_1 = require('./actions/test/getTxpoolStatus.js')
Object.defineProperty(exports, 'getTxpoolStatus', {
	enumerable: true,
	get: function () {
		return getTxpoolStatus_js_1.getTxpoolStatus
	},
})
const impersonateAccount_js_1 = require('./actions/test/impersonateAccount.js')
Object.defineProperty(exports, 'impersonateAccount', {
	enumerable: true,
	get: function () {
		return impersonateAccount_js_1.impersonateAccount
	},
})
const increaseTime_js_1 = require('./actions/test/increaseTime.js')
Object.defineProperty(exports, 'increaseTime', {
	enumerable: true,
	get: function () {
		return increaseTime_js_1.increaseTime
	},
})
const inspectTxpool_js_1 = require('./actions/test/inspectTxpool.js')
Object.defineProperty(exports, 'inspectTxpool', {
	enumerable: true,
	get: function () {
		return inspectTxpool_js_1.inspectTxpool
	},
})
const mine_js_1 = require('./actions/test/mine.js')
Object.defineProperty(exports, 'mine', {
	enumerable: true,
	get: function () {
		return mine_js_1.mine
	},
})
const removeBlockTimestampInterval_js_1 = require('./actions/test/removeBlockTimestampInterval.js')
Object.defineProperty(exports, 'removeBlockTimestampInterval', {
	enumerable: true,
	get: function () {
		return removeBlockTimestampInterval_js_1.removeBlockTimestampInterval
	},
})
const reset_js_1 = require('./actions/test/reset.js')
Object.defineProperty(exports, 'reset', {
	enumerable: true,
	get: function () {
		return reset_js_1.reset
	},
})
const revert_js_1 = require('./actions/test/revert.js')
Object.defineProperty(exports, 'revert', {
	enumerable: true,
	get: function () {
		return revert_js_1.revert
	},
})
const sendUnsignedTransaction_js_1 = require('./actions/test/sendUnsignedTransaction.js')
Object.defineProperty(exports, 'sendUnsignedTransaction', {
	enumerable: true,
	get: function () {
		return sendUnsignedTransaction_js_1.sendUnsignedTransaction
	},
})
const setAutomine_js_1 = require('./actions/test/setAutomine.js')
Object.defineProperty(exports, 'setAutomine', {
	enumerable: true,
	get: function () {
		return setAutomine_js_1.setAutomine
	},
})
const setBalance_js_1 = require('./actions/test/setBalance.js')
Object.defineProperty(exports, 'setBalance', {
	enumerable: true,
	get: function () {
		return setBalance_js_1.setBalance
	},
})
const setBlockGasLimit_js_1 = require('./actions/test/setBlockGasLimit.js')
Object.defineProperty(exports, 'setBlockGasLimit', {
	enumerable: true,
	get: function () {
		return setBlockGasLimit_js_1.setBlockGasLimit
	},
})
const setBlockTimestampInterval_js_1 = require('./actions/test/setBlockTimestampInterval.js')
Object.defineProperty(exports, 'setBlockTimestampInterval', {
	enumerable: true,
	get: function () {
		return setBlockTimestampInterval_js_1.setBlockTimestampInterval
	},
})
const setCode_js_1 = require('./actions/test/setCode.js')
Object.defineProperty(exports, 'setCode', {
	enumerable: true,
	get: function () {
		return setCode_js_1.setCode
	},
})
const setCoinbase_js_1 = require('./actions/test/setCoinbase.js')
Object.defineProperty(exports, 'setCoinbase', {
	enumerable: true,
	get: function () {
		return setCoinbase_js_1.setCoinbase
	},
})
const setIntervalMining_js_1 = require('./actions/test/setIntervalMining.js')
Object.defineProperty(exports, 'setIntervalMining', {
	enumerable: true,
	get: function () {
		return setIntervalMining_js_1.setIntervalMining
	},
})
const setLoggingEnabled_js_1 = require('./actions/test/setLoggingEnabled.js')
Object.defineProperty(exports, 'setLoggingEnabled', {
	enumerable: true,
	get: function () {
		return setLoggingEnabled_js_1.setLoggingEnabled
	},
})
const setMinGasPrice_js_1 = require('./actions/test/setMinGasPrice.js')
Object.defineProperty(exports, 'setMinGasPrice', {
	enumerable: true,
	get: function () {
		return setMinGasPrice_js_1.setMinGasPrice
	},
})
const setNextBlockBaseFeePerGas_js_1 = require('./actions/test/setNextBlockBaseFeePerGas.js')
Object.defineProperty(exports, 'setNextBlockBaseFeePerGas', {
	enumerable: true,
	get: function () {
		return setNextBlockBaseFeePerGas_js_1.setNextBlockBaseFeePerGas
	},
})
const setNextBlockTimestamp_js_1 = require('./actions/test/setNextBlockTimestamp.js')
Object.defineProperty(exports, 'setNextBlockTimestamp', {
	enumerable: true,
	get: function () {
		return setNextBlockTimestamp_js_1.setNextBlockTimestamp
	},
})
const setNonce_js_1 = require('./actions/test/setNonce.js')
Object.defineProperty(exports, 'setNonce', {
	enumerable: true,
	get: function () {
		return setNonce_js_1.setNonce
	},
})
const setStorageAt_js_1 = require('./actions/test/setStorageAt.js')
Object.defineProperty(exports, 'setStorageAt', {
	enumerable: true,
	get: function () {
		return setStorageAt_js_1.setStorageAt
	},
})
const snapshot_js_1 = require('./actions/test/snapshot.js')
Object.defineProperty(exports, 'snapshot', {
	enumerable: true,
	get: function () {
		return snapshot_js_1.snapshot
	},
})
const setRpcUrl_js_1 = require('./actions/test/setRpcUrl.js')
Object.defineProperty(exports, 'setRpcUrl', {
	enumerable: true,
	get: function () {
		return setRpcUrl_js_1.setRpcUrl
	},
})
const stopImpersonatingAccount_js_1 = require('./actions/test/stopImpersonatingAccount.js')
Object.defineProperty(exports, 'stopImpersonatingAccount', {
	enumerable: true,
	get: function () {
		return stopImpersonatingAccount_js_1.stopImpersonatingAccount
	},
})
//# sourceMappingURL=test.js.map
