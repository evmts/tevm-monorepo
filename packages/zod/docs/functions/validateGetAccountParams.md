**@tevm/zod** ∙ [README](../README.md) ∙ [API](../API.md)

***

[API](../API.md) > validateGetAccountParams

# Function: validateGetAccountParams()

> **validateGetAccountParams**(`action`): `GetAccountError`[]

## Parameters

▪ **action**: `GetAccountParams`

## Returns

## Source

[packages/zod/src/validators/validateGetAccountParams.js:8](https://github.com/evmts/tevm-monorepo/blob/main/packages/zod/src/validators/validateGetAccountParams.js#L8)

***
Generated using [typedoc-plugin-markdown](https://www.npmjs.com/package/typedoc-plugin-markdown) and [TypeDoc](https://typedoc.org/)
