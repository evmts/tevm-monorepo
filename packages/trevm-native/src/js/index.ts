import { createLogger } from '@tevm/logger';
import { InternalError } from '@tevm/errors';
import type { StateManager } from '@tevm/state';
import type { Common } from '@tevm/common';
import type { Chain } from '@tevm/blockchain';

// Import native module
import { init_trevm, TrevmWrapper } from '../../dist/index.node';

export interface TrevmEvmOptions {
  common: Common;
  stateManager: StateManager;
  blockchain: Chain;
  customPrecompiles?: any[];
  profiler?: boolean;
  allowUnlimitedContractSize?: boolean;
  loggingLevel?: string;
}

export interface TrevmRunCallOptions {
  caller?: string;
  to?: string;
  data?: string;
  value?: bigint;
  gasLimit?: bigint;
  skipBalance?: boolean;
}

export interface TrevmRunCallResult {
  returnValue: string;
  executionGasUsed: string;
  error?: string;
}

export interface TrevmEvm {
  runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult>;
}

/**
 * Creates a trevm-based EVM instance
 * @param options EVM options
 * @returns A trevm EVM instance
 */
export const createTrevm = async ({
  common,
  stateManager,
  blockchain,
  customPrecompiles,
  profiler,
  allowUnlimitedContractSize,
  loggingLevel,
}: TrevmEvmOptions): Promise<TrevmEvm> => {
  const logger = createLogger({
    name: '@tevm/trevm',
    level: loggingLevel ?? 'warn',
  });
  
  logger.debug({
    allowUnlimitedContractSize,
    profiler,
    customPrecompiles: customPrecompiles?.map(c => c.address.toString()),
  });

  try {
    // Initialize the native trevm module with callbacks to the StateManager
    const nativeTrevm = await init_trevm({
      chainId: common.chainId(),
      allowUnlimitedContractSize: allowUnlimitedContractSize ?? false,
      stateCallbacks: {
        getAccount: async (address: string) => {
          try {
            const account = await stateManager.getAccount(address);
            if (!account) return null;
            
            return {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? `0x${Buffer.from(account.codeHash).toString('hex')}` : undefined,
              code: account.code ? `0x${Buffer.from(account.code).toString('hex')}` : undefined,
            };
          } catch (error) {
            logger.error(`Error in getAccount for ${address}:`, error);
            return null;
          }
        },
        putAccount: async (address: string, account: any) => {
          try {
            await stateManager.putAccount(address, {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? Buffer.from(account.codeHash.slice(2), 'hex') : undefined,
              code: account.code ? Buffer.from(account.code.slice(2), 'hex') : undefined,
            });
            return true;
          } catch (error) {
            logger.error(`Error in putAccount for ${address}:`, error);
            return false;
          }
        },
        getStorage: async (address: string, key: string) => {
          try {
            const value = await stateManager.getContractStorage(address, key);
            return value ? `0x${Buffer.from(value).toString('hex')}` : '0x0';
          } catch (error) {
            logger.error(`Error in getStorage for ${address} key ${key}:`, error);
            return '0x0';
          }
        },
        putStorage: async (address: string, key: string, value: string) => {
          try {
            await stateManager.putContractStorage(
              address,
              key,
              Buffer.from(value.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putStorage for ${address} key ${key}:`, error);
            return false;
          }
        },
        getCode: async (address: string) => {
          try {
            const code = await stateManager.getContractCode(address);
            return code ? `0x${Buffer.from(code).toString('hex')}` : '0x';
          } catch (error) {
            logger.error(`Error in getCode for ${address}:`, error);
            return '0x';
          }
        },
        putCode: async (address: string, code: string) => {
          try {
            await stateManager.putContractCode(
              address,
              Buffer.from(code.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putCode for ${address}:`, error);
            return false;
          }
        },
        getBlockHash: async (blockNumber: number) => {
          try {
            const block = await blockchain.getBlock(BigInt(blockNumber));
            return block ? `0x${Buffer.from(block.header.hash()).toString('hex')}` : null;
          } catch (error) {
            logger.error(`Error in getBlockHash for block ${blockNumber}:`, error);
            return null;
          }
        }
      }
    });

    // Return a JavaScript wrapper around the native trevm that matches the Evm interface
    return {
      async runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult> {
        try {
          return await nativeTrevm.run_call({
            caller: opts.caller,
            to: opts.to,
            data: opts.data,
            value: opts.value,
            gas_limit: opts.gasLimit,
            skip_balance: opts.skipBalance,
          });
        } catch (error) {
          logger.error('Error in runCall:', error);
          throw new InternalError(`Trevm execution error: ${error}`);
        }
      }
    };
  } catch (error) {
    logger.error('Error initializing trevm:', error);
    throw new InternalError(`Failed to initialize trevm: ${error}`);
  }
};