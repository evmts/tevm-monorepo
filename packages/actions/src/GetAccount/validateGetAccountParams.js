import { InvalidRequestError } from '@tevm/errors'
import { zGetAccountParams } from './zGetAccountParams.js'

/**
 * @typedef {InvalidRequestError} ValidateGetAccountParamsError
 */

/**
 * @param {import('./GetAccountParams.js').GetAccountParams} action
 * @returns {Array<ValidateGetAccountParamsError>}
 */
export const validateGetAccountParams = (action) => {
	/**
	 * @type {Array<ValidateGetAccountParamsError>}
	 */
	const errors = []

	const parsedParams = zGetAccountParams.safeParse(action)

	if (parsedParams.success === false) {
		const formattedErrors = parsedParams.error.format()
		formattedErrors._errors.forEach((error) => {
			errors.push(new InvalidRequestError(error))
		})
	}

	return errors
}
