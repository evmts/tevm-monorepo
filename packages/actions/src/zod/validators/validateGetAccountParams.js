import { InvalidRequestError } from '@tevm/errors'
import { zGetAccountParams } from '../params/index.js'

/**
 * @typedef {InvalidRequestError} ValidateGetAccountParamsError
 */

/**
 * @param {import('@tevm/actions').GetAccountParams} action
 * @returns {Array<ValidateGetAccountParamsError>}
 */
export const validateGetAccountParams = (action) => {
	/**
	 * @type {Array<ValidateGetAccountParamsError>}
	 */
	const errors = []

	const parsedParams = zGetAccountParams.safeParse(action)

	if (parsedParams.success === false) {
		const formattedErrors = parsedParams.error.format()
		formattedErrors._errors.forEach((error) => {
			errors.push(new InvalidRequestError(error))
		})
	}

	return errors
}
