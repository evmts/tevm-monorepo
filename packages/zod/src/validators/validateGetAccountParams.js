import { createError } from '../common/index.js'
import { zGetAccountParams } from '../params/index.js'

/**
 * @param {import('@tevm/actions-types').GetAccountParams} action
 * @returns {Array<import('@tevm/errors').GetAccountError>}
 */
export const validateGetAccountParams = (action) => {
	/**
	 * @type {Array<import('@tevm/errors').GetAccountError>}
	 */
	const errors = []

	const parsedParams = zGetAccountParams.safeParse(action)

	if (parsedParams.success === false) {
		const formattedErrors = parsedParams.error.format()
		formattedErrors._errors.forEach((error) => {
			errors.push(createError('InvalidRequestError', error, String(action)))
		})
	}

	return errors
}
