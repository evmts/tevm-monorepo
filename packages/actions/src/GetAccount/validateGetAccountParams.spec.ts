import { describe, expect, it } from 'bun:test'
import { InvalidAddressError, InvalidRequestError } from '@tevm/errors'
import { validateGetAccountParams } from './validateGetAccountParams.js'

describe('validateGetAccountParams', () => {
	it('should return an empty array if parameters are valid', () => {
		const validParams = {
			address: `0x${'1'.repeat(40)}`,
			throwOnFail: true,
			returnStorage: true,
		} as const
		const errors = validateGetAccountParams(validParams)
		expect(errors).toEqual([])
	})

	it('should return InvalidAddressError for an invalid address', () => {
		const invalidParams = {
			address: 'invalidAddress',
			throwOnFail: true,
			returnStorage: true,
		}
		const errors = validateGetAccountParams(invalidParams as any)
		expect(errors).toHaveLength(1)
		expect(errors[0]).toBeInstanceOf(InvalidAddressError)
		expect(errors).toMatchSnapshot()
	})

	it('should return InvalidRequestError for an invalid throwOnFail parameter', () => {
		const invalidParams = {
			address: `0x${'1'.repeat(40)}`,
			throwOnFail: 'notABoolean',
			returnStorage: true,
		}
		const errors = validateGetAccountParams(invalidParams as any)
		expect(errors).toHaveLength(1)
		expect(errors[0]).toBeInstanceOf(InvalidRequestError)
		expect(errors).toMatchSnapshot()
	})

	it('should return InvalidRequestError for an invalid returnStorage parameter', () => {
		const invalidParams = {
			address: `0x${'1'.repeat(40)}`,
			throwOnFail: true,
			returnStorage: 'notABoolean',
		}
		const errors = validateGetAccountParams(invalidParams as any)
		expect(errors).toHaveLength(1)
		expect(errors[0]).toBeInstanceOf(InvalidRequestError)
		expect(errors).toMatchSnapshot()
	})

	it('should return multiple errors for multiple invalid parameters', () => {
		const invalidParams = {
			address: 'invalidAddress',
			throwOnFail: 'notABoolean',
			returnStorage: 'notABoolean',
		}
		const errors = validateGetAccountParams(invalidParams as any)
		expect(errors).toHaveLength(3)
		expect(errors).toMatchSnapshot()
	})

	it('should handle additional errors in the formattedErrors._errors array', () => {
		const invalidParams = {
			address: `0x${'1'.repeat(40)}`,
			throwOnFail: true,
			returnStorage: true,
		}
		const errors = validateGetAccountParams(invalidParams as any)
		// Manually adding an error to the _errors array for testing
		errors.push(new InvalidRequestError('Test error'))
		expect(errors).toHaveLength(1)
		expect(errors[0]).toBeInstanceOf(InvalidRequestError)
		expect(errors).toMatchSnapshot()
	})
})
