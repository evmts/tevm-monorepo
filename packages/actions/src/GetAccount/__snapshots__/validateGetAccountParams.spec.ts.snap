// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`validateGetAccountParams should return InvalidAddressError for an invalid address 1`] = `
[
  [InvalidAddressError: Invalid address param. Invalid Address invalidAddress

Docs: https://tevm.sh/reference/tevm/errors/classes/invalidaddresserror/
Version: 1.1.0.next-73],
]
`;

exports[`validateGetAccountParams should return InvalidRequestError for an invalid throwOnFail parameter 1`] = `
[
  [InvalidRequestError: Invalid throwOnFail param. throwOnFail must be a boolean or not provided. Expected boolean, received string

Docs: https://tevm.sh/reference/tevm/errors/classes/invalidrequesterror/
Version: 1.1.0.next-73],
]
`;

exports[`validateGetAccountParams should return InvalidRequestError for an invalid returnStorage parameter 1`] = `
[
  [InvalidRequestError: Invalid returnStorage param. returnStorage must be a boolean or not provided. Expected boolean, received string

Docs: https://tevm.sh/reference/tevm/errors/classes/invalidrequesterror/
Version: 1.1.0.next-73],
]
`;

exports[`validateGetAccountParams should handle additional errors in the formattedErrors._errors array 1`] = `
[
  [InvalidRequestError: Test error

Docs: https://tevm.sh/reference/tevm/errors/classes/invalidrequesterror/
Version: 1.1.0.next-73],
]
`;

exports[`validateGetAccountParams should return multiple errors for multiple invalid parameters 1`] = `
[
  [InvalidRequestError: Invalid throwOnFail param. throwOnFail must be a boolean or not provided. Expected boolean, received string

Docs: https://tevm.sh/reference/tevm/errors/classes/invalidrequesterror/
Version: 1.1.0.next-73],
  [InvalidRequestError: Invalid returnStorage param. returnStorage must be a boolean or not provided. Expected boolean, received string

Docs: https://tevm.sh/reference/tevm/errors/classes/invalidrequesterror/
Version: 1.1.0.next-73],
  [InvalidAddressError: Invalid address param. Invalid Address invalidAddress

Docs: https://tevm.sh/reference/tevm/errors/classes/invalidaddresserror/
Version: 1.1.0.next-73],
]
`;
